### PR TITLE
standalone  hypercontraction module

### PR DESCRIPTION
**Pull request by liuzhenhai93**
- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

* 1、move hyper connection contraction to Its own module
* 2、merge alpha_pre / alpha_post / alpha_res into unified scale for better HF checkpoint alignment

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

https://github.com/NVIDIA/Megatron-LM/issues/5670

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/hyper_connection.py [hunk 1]
```diff
@@ -195,10 +195,9 @@ def __init__(self, config: TransformerConfig, layer_number: int):
         )
 
         init_alpha = config.mhc_init_gating_factor
-        # Learnable scaling factors (Eq. 5 in paper)
-        self.alpha_pre = nn.Parameter(torch.full((1,), init_alpha))
-        self.alpha_post = nn.Parameter(torch.full((1,), init_alpha))
-        self.alpha_res = nn.Parameter(torch.full((1,), init_alpha))
+        # Learnable scaling factors (Eq. 5 in paper): pre, post, res in one (3,) tensor
+        # matching HF checkpoint layout (attn_hc.scale / ffn_hc.scale).
+        self.scale = nn.Parameter(torch.full((3,), init_alpha))
```
Comment (cspades) [inline megatron/core/transformer/hyper_connection.py:200]: This will affect backwards compatibility with earlier checkpoints right?

Can you handle the HF parameter format online during checkpoint loading?

File: megatron/core/transformer/multi_token_prediction.py [hunk 2]
```diff
@@ -1235,16 +1235,7 @@ def __init__(
...
-            hc_dim = self.config.hidden_size * hc_mult
-            self.hc_head_fn = nn.Parameter(torch.randn(hc_mult, hc_dim))
-            self.hc_head_base = nn.Parameter(torch.zeros(hc_mult))
-            self.hc_head_scale = nn.Parameter(torch.ones(1))
-            nn.init.xavier_uniform_(self.hc_head_fn)
-            if self.config.sequence_parallel:
-                setattr(self.hc_head_fn, 'sequence_parallel', True)
-                setattr(self.hc_head_base, 'sequence_parallel', True)
-                setattr(self.hc_head_scale, 'sequence_parallel', True)
+            self.mhc_contract = HyperConnectionContractModule(self.config)
```
[the remark this answers is not in the document]
Comment (cspades) [reply megatron/core/transformer/multi_token_prediction.py:1238]: Yup, pretty much my concern as well, we need to make sure that checkpointing works as a requirement to performing this refactor.

**Review by cspades — COMMENTED** [review]
Needs backwards compatibility at the very least, to map existing checkpoints FQN's to the `mhc_contract` parameter data. Do agree this is DRY though. Checkpointing can be messier if it means the code can be reused.
